fqn_decorators/asynchronous.py

```diff
@@ -9,19 +9,27 @@
 
 
 class AsyncDecorator(Decorator):
-    async def __call__(self, *args, **kwargs):
-        self.fqn = self.get_fqn()
-        self.args = args
-        self.kwargs = kwargs
+    def __new__(cls, func=None, **params):
+        def sync_call(self, *args, **kwargs):
+            return self.__class__(args[0], **self.params)
 
-        self.before()
-        try:
-            self.result = await self.func(*self.args, **self.kwargs)
-        except:
-            self.exc_info = sys.exc_info()
-            self.exception()
-            raise
-        finally:
-            self.after()
+        async def async_call(self, *args, **kwargs):
```

Comment:
I think it should be defined neither on the class nor outside (as a function) as it's internal-only and is created only as part of `__new__` call.

Reply:
but if it uses `self`, I would say it should be defined on class level, isn't it?

Reply:
also it will create 2 new functions every time new instance of this class is being created, which is meaningless(though it should not be harmful).

the same applies to `_wrap_func` as well.

Reply:
@kammala the purpose of this pull request is to dynamically create class methods upon object creation. I.e. when this decorator object is created without parameters its `__call__` should be async and handle its `self.func` while when first the decorator is used with parameters its `__call__` method should be synchronous.

That's why I specifically didn't move these two internal methods on a class-level as only one of them should propagate to a real `self.__call__()` method upon decorator class object creation while the other one should not exist at all.